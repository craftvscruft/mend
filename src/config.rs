use std::path::Path;
use std::fs;
use std::collections::BTreeMap;
use anyhow::{anyhow, Context};
use crate::{Cli, Mend};

pub fn load_mend(cli: &Cli) ->anyhow::Result<Mend> {
    let parent_dir = Path::new(&cli.file)
        .parent()
        .expect("Unable to get the parent directory");


    let contents = match fs::read_to_string(&cli.file) {
        Ok(c) => c,
        Err(e) => {
            eprintln!("Could not read file `{}` {}", &cli.file, e);
            panic!()
        }
    };

    let main_mend: Mend = toml::from_str(&contents)
    .with_context(|| format!("Unable to load data from `{}`", &cli.file))?;

    let mut merged_mend: Mend = Mend {
        from: None,
        include: Vec::new(),
        env: BTreeMap::new(),
        recipes: BTreeMap::new(),
        hooks: BTreeMap::new(),
        steps: Vec::new(),
    };
    for include_file in &main_mend.include {
        let include_contents = fs::read_to_string(parent_dir.join(include_file))
            .with_context(|| format!("Could not read include file `{}`", &include_file))?;
        let include_mend: Mend = toml::from_str(&include_contents)
            .with_context(|| format!("Unable to load data from `{}`", &include_file))?;
        if !include_mend.steps.is_empty() {
            return Err(anyhow!("We only allow includes 1 level deep, sorry. Please restructure `{}`", &include_file))
        }
        crate::extend_mend(&mut merged_mend, include_mend);
    }
    crate::extend_mend(&mut merged_mend, main_mend);
    for recipe_entry in merged_mend.recipes.values_mut() {
        match recipe_entry.tag.take() {
            Some(tag) => {
                recipe_entry.tags.push(tag);
                recipe_entry.tag = None
            }
            None => {}
        }
    }
    Ok(merged_mend)
}
