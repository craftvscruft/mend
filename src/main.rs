use anyhow::{anyhow, Context};
use clap::Parser;
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;
use std::fmt::Debug;
use std::fs;
use std::path::Path;
use std::process::exit;

mod config;

#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
pub struct Cli {
    #[arg(short='f', long="file", default_value_t=String::from("mend.toml"))]
    file: String,
}
#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct Mend {
    from: Option<From>,

    #[serde(default)]
    include: Vec<String>,

    #[serde(default)]
    env: BTreeMap<String, String>,

    #[serde(default)]
    recipes: BTreeMap<String, Recipe>,

    #[serde(default)]
    hooks: BTreeMap<String, Hook>,

    #[serde(default)]
    steps: Vec<String>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct From {
    sha: String,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct Recipe {
    run: String,
    commit_template: Option<String>,
    tag: Option<String>,

    #[serde(default)]
    tags: Vec<String>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct Hook {
    run: Option<String>,
    run_for_tag: Option<HookRunForTag>,
}

#[derive(Debug, PartialEq, Serialize, Deserialize)]
pub struct HookRunForTag {
    tag: String,
    run: String,
    else_run: Option<String>,
}

fn main() {
    run(&Cli::parse());
}

fn run(cli: &Cli) {
    match config::load_mend(cli) {
        Ok(merged_mend) => {
            match toml::to_string_pretty(&merged_mend) {
                Ok(text) => {
                    println!("{}", text);
                }
                Err(e) => {
                    eprintln!("{e}");
                    exit(1);
                }
            }
        }
        Err(e) => {
            eprintln!("{e}");
            exit(1);
        }
    }

}

fn extend_mend(merged_mend: &mut Mend, include_mend: Mend) {
    merged_mend.env.extend(include_mend.env);
    merged_mend.from = include_mend.from;
    merged_mend.recipes.extend(include_mend.recipes);
    merged_mend.hooks.extend(include_mend.hooks);
    for ele in include_mend.steps {
        merged_mend.steps.push(ele)
    }
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use crate::Cli;
    use crate::config::load_mend;

    #[test]
    fn it_works() {
        let mut toml_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        toml_path.push("examples/mend.toml");
        let args = Cli {
            file: String::from(toml_path.to_str().unwrap())
        };
        let loaded = load_mend(&args);
        insta::assert_yaml_snapshot!(loaded.expect("Failed loading"));
    }
}
