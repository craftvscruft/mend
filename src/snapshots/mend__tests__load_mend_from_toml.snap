---
source: src/main.rs
expression: "loaded.expect(\"Failed loading\")"
---
from:
  sha: 43a3a253
  repo: ~/dev/ioccc/endoh2
include: []
env:
  DEFAULT_FILE: main.c
  JAVA_HOME: /Library/Java/JavaVirtualMachines/graalvm-jdk-20.0.2+9.1/Contents/Home/
  PATH: "$PATH:/Users/rmyers/dev/untangler/build/install/untangler/bin"
recipes:
  format:
    run: clang-format -i $DEFAULT_FILE
    commit_template: d - Format
    tag: ~
    tags:
      - binary_identical
  move_includes_to_top:
    run: "    grep \"^#include\" $DEFAULT_FILE > a.tmp && grep -v \"^#include\" $DEFAULT_FILE >> a.tmp && mv a.tmp $DEFAULT_FILE\n  "
    commit_template: r - Move includes to top
    tag: ~
    tags:
      - binary_identical
  remove_comments:
    run: "    untangler remove comment \"*\" --sub=\" \" -w -f $DEFAULT_FILE\n  "
    commit_template: d - Remove comments
    tag: ~
    tags:
      - binary_identical
  remove_comments_in_includes:
    run: "    perl -pi -e 's{^#include */\\*((?!\\*/).)*\\*/}{#include}gs' $DEFAULT_FILE\n  "
    commit_template: d - Remove comments in includes
    tag: ~
    tags:
      - binary_identical
  rename:
    run: untangler rename $1 $2 -w -f $DEFAULT_FILE
    commit_template: R - Rename $1 to $2
    tag: ~
    tags: []
  split_declarations:
    run: "    untangler misc split-declaration \"*\" -w -f $DEFAULT_FILE\n  "
    commit_template: r - Split declarations
    tag: ~
    tags:
      - binary_identical
hooks:
  after_step:
    - run: diff a.out a.out.bak
      when_tag: binary_identical
      when_not_tag: ~
    - run: make test
      when_tag: ~
      when_not_tag: binary_identical
  before_step:
    - run: make && cp a.out a.out.bak
      when_tag: binary_identical
      when_not_tag: ~
    - run: make
      when_tag: ~
      when_not_tag: binary_identical
steps:
  - remove_comments_in_includes
  - remove_comments
  - format
  - move_includes_to_top
  - split_declarations
  - rename B calculate_value
  - rename b evaluate_and_draw_interval
  - rename X control_points
  - rename P point_value
  - rename O offset
  - rename m pixel_index
  - rename k color_value
  - rename S screen_buffer

